model/src/gateway/event/mod.rs

```diff
@@ -267,3 +269,35 @@ impl From<ShardEvent> for Event {
         }
     }
 }
+
+/// An error that describes a failure to convert
+/// from one event type to another.
+#[derive(Debug)]
+pub struct EventConversionError<T> {
+    /// The original event
+    event: T,
+    msg: &'static str,
```

Comment:
We probably don't need this message, the Display formatter can just always be "event variant failed to convert", it will be obvious from the user's code path what it was being converted into